### PR TITLE
Feature/402 atna composition endpoint

### PR DESCRIPTION
**Pull request by subigre**
## Changes

Implement ATNA profile in Composition endpoint.

- Introduce new `audit` package in `rest-openehr` module.
- Create some Spring HandlerInterceptors to populate the AuditDataset based on `HttpServletRequest` and `HttpServletResponse`
- Resgister interceptor in the `WebMvcConfig.java` if audit is enabled.

## Related issue

Resolves ehrbase/project_management#402

### CODE REVIEW

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrEventIdCode.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package org.ehrbase.rest.openehr.audit;
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrEventIdCode.java:1]: Please add a license header to new files.

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrEventTypeCode.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package org.ehrbase.rest.openehr.audit;
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrEventTypeCode.java:1]: Please add a license header to new files.

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrAuditStrategy.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
+package org.ehrbase.rest.openehr.audit;
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrAuditStrategy.java:1]: Please add a license header to new files.

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/support/OpenEhrAuditMessageBuilder.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
+package org.ehrbase.rest.openehr.audit.support;
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/support/OpenEhrAuditMessageBuilder.java:1]: Please add a license header to new files.

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/support/CompositionAuditMessageBuilder.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
+package org.ehrbase.rest.openehr.audit.support;
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/support/CompositionAuditMessageBuilder.java:1]: Please add a license header to new files.

**Review by jakesmolka — CHANGES_REQUESTED** [review]
Your changes are looking good! It took me moment to understand how the IPF ATNA auditing works, but I got it now. 
I have made several comments where you should add the license header (see any other source file) and I commented two questions.
And please add a line to the changelog.

Apart from that: What local dev setup do I need to test everything? I think a small explanation in our documentation repo would be great!

And currently the test are failing. After a quick look it looks like your changes to default basic auth might be the problem. But it was only a quick look, so I'm not sure.

File: application/src/main/resources/application-local.yml [hunk 6]
```diff
@@ -39,7 +39,9 @@ server:
     iterationScanDepth: 2
 
 security:
-  authType: NONE
+  authType: basic
+  auth-user: myuser
+  auth-password: myPassword432
```
[the remark this answers is not in the document]
Comment (subigre) [reply application/src/main/resources/application-local.yml:44]: I reverted this change. 
My mistake, I wanted to be sure that I am able to retrieve the principal from the HttpServletRequest and I pushed the changes that I made for testing purposes.

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/CompositionAuditDataset.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
+package org.ehrbase.rest.openehr.audit;
```
[the remark this answers is not in the document]
Comment (subigre) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/CompositionAuditDataset.java:1]: I forgot this part but the files are now updated

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/CompositionAuditStrategy.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
+package org.ehrbase.rest.openehr.audit;
```
[the remark this answers is not in the document]
Comment (subigre) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/CompositionAuditStrategy.java:1]: Done

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrAuditDataset.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
+package org.ehrbase.rest.openehr.audit;
```
[the remark this answers is not in the document]
Comment (subigre) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrAuditDataset.java:1]: Done

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrEventIdCode.java (hunk 1 above)
Comment (subigre) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrEventIdCode.java:1]: Done

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrEventTypeCode.java (hunk 2 above)
Comment (subigre) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrEventTypeCode.java:1]: Done

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrAuditStrategy.java (hunk 3 above)
Comment (subigre) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/OpenEhrAuditStrategy.java:19]: Done

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/support/CompositionAuditMessageBuilder.java (hunk 5 above)
Comment (subigre) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/support/CompositionAuditMessageBuilder.java:19]: Done

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/support/OpenEhrAuditMessageBuilder.java (hunk 4 above)
Comment (subigre) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/audit/support/OpenEhrAuditMessageBuilder.java:19]: Done

File: application/src/main/resources/application.yml [hunk 10]
```diff
@@ -138,4 +138,11 @@ management:
...
-        enabled: true
\ No newline at end of file
+        enabled: true
+
+# Audit Properties
+ipf:
+  atna:
+    audit-enabled: false
+    audit-repository-host:
+    audit-repository-port:
```
[the remark this answers is not in the document]
Comment (subigre) [reply application/src/main/resources/application.yml:148]: Here is an example of my local configuration:
```json
ipf:
  atna:
    audit-enabled: true
    audit-repository-host: localhost         # Hostname of the audit repository
    audit-repository-port: 514               # Port of the audit repository (UDP by default)
```

**Comment by subigre** [comment]
Everything is implemented here https://github.com/ehrbase/ehrbase/pull/452

**Pull request closed without merging**